File: frontend/src/api/proxy.ts
import apiClient from './client'

// Proxy Manager API — the egress proxy pool (add / switch / health-check) and the
// outbound flow log. Mirrors the backend routes under /system/proxies and
// /system/proxy/flows.

export type ProxyLane = 'default' | 'osint' | 'vulnscan'

export interface ProxyProfile {
  id: number
  name: string
  url: string // masked (credentials stripped) — server never returns the raw URL
  no_proxy: string
  fallback_direct: boolean
  is_active: boolean
  lane: ProxyLane
  quota_bytes: number
  quota_hard_stop: boolean
  quota_used_bytes: number
  over_quota: boolean
  healthy: boolean
  latency_ms: number
  last_error: string
  last_check: string | null
  exit_ip: string
  exit_country: string
  exit_org: string
  is_tor: boolean
  exit_checked_at: string | null
  exit_ip_prev: string
  identity_drift: boolean
  created_at: string
  updated_at: string
}

export interface ProxyFlow {
  id: number
  created_at: string
  proxy_label: string
  via_proxy: boolean
  leaked: boolean
  source: string
  method: string
  scheme: string
  host: string
  url: string
  status: number
  content_type: string
  tls_version: string
  bytes_out: number
  bytes_in: number
  duration_ms: number
  dns_ms: number
  connect_ms: number
  tls_ms: number
  ttfb_ms: number
  error: string
}

export interface ProxyFlowStats {
  count: number
  bytes_in: number
  bytes_out: number
  errors: number
  proxied: number
  direct: number
  leaked: number
  coverage_pct: number
  by_proxy: Record<string, number>
}

export interface ProxyPoolMode {
  id: number
  mode: 'manual' | 'failover' | 'rotate'
  interval_sec: number
  kill_switch: boolean
}

export interface ProxyHealthHistory {
  samples: { id: number; profile_id: number; healthy: boolean; latency_ms: number; created_at: string }[]
  uptime_pct: number
  count: number
}

export interface ProxyLeakTest {
  exit_ips: Record<string, string>
  exit_ip: string
  consistent: boolean
  checked_at: string
  error?: string
}

export interface ProxyAnalytics {
  since_hours: number
  total: number
  proxied: number
  leaked: number
  coverage_pct: number
  top_hosts: { host: string; count: number; bytes: number }[]
  per_proxy: { proxy_label: string; count: number; errors: number; bytes_in: number; bytes_out: number; avg_ms: number; p50_ms: number; p95_ms: number }[]
}

export interface ProxyProfilePayload {
  name: string
  url: string
  no_proxy?: string
  fallback_direct?: boolean
  lane?: ProxyLane
  quota_bytes?: number
  quota_hard_stop?: boolean
}

interface ApiResponse<T> {
  success: boolean
  data: T
  source?: string
}

export const proxyApi = {
  list: async (): Promise<ProxyProfile[]> => {
    const { data } = await apiClient.get<ApiResponse<ProxyProfile[]>>('/system/proxies')
    return data.data
  },

  create: async (body: ProxyProfilePayload): Promise<ProxyProfile> => {
    const { data } = await apiClient.post<ApiResponse<ProxyProfile>>('/system/proxies', body)
    return data.data
  },

  update: async (id: number, body: ProxyProfilePayload): Promise<ProxyProfile> => {
    const { data } = await apiClient.patch<ApiResponse<ProxyProfile>>(`/system/proxies/${id}`, body)
    return data.data
  },

  remove: async (id: number): Promise<void> => {
    await apiClient.delete(`/system/proxies/${id}`)
  },

  activate: async (id: number): Promise<ProxyProfile> => {
    const { data } = await apiClient.post<ApiResponse<ProxyProfile>>(`/system/proxies/${id}/activate`)
    return data.data
  },

  deactivate: async (): Promise<void> => {
    await apiClient.post('/system/proxies/deactivate')
  },

  check: async (id: number): Promise<ProxyProfile> => {
    const { data } = await apiClient.post<ApiResponse<ProxyProfile>>(`/system/proxies/${id}/check`)
    return data.data
  },

  identity: async (id: number): Promise<ProxyProfile> => {
    const { data } = await apiClient.post<ApiResponse<ProxyProfile>>(`/system/proxies/${id}/identity`)
    return data.data
  },

  getMode: async (): Promise<ProxyPoolMode> => {
    const { data } = await apiClient.get<ApiResponse<ProxyPoolMode>>('/system/proxies/mode')
    return data.data
  },

  setMode: async (body: { mode?: string; interval_sec?: number; kill_switch?: boolean }): Promise<ProxyPoolMode> => {
    const { data } = await apiClient.post<ApiResponse<ProxyPoolMode>>('/system/proxies/mode', body)
    return data.data
  },

  bulkCreate: async (text: string, lane: ProxyLane): Promise<{ created: number; errors: string[] }> => {
    const { data } = await apiClient.post<ApiResponse<{ created: number; errors: string[] }>>('/system/proxies/bulk', { text, lane })
    return data.data
  },

  checkAll: async (): Promise<ProxyProfile[]> => {
    const { data } = await apiClient.post<ApiResponse<ProxyProfile[]>>('/system/proxies/check-all')
    return data.data
  },

  healthHistory: async (id: number, hours = 24): Promise<ProxyHealthHistory> => {
    const { data } = await apiClient.get<ApiResponse<ProxyHealthHistory>>(`/system/proxies/${id}/health-history?hours=${hours}`)
    return data.data
  },

  leakTest: async (id: number): Promise<ProxyLeakTest> => {
    const { data } = await apiClient.post<ApiResponse<ProxyLeakTest>>(`/system/proxies/${id}/leak-test`)
    return data.data
  },

  analytics: async (sinceHours = 24): Promise<ProxyAnalytics> => {
    const { data } = await apiClient.get<ApiResponse<ProxyAnalytics>>(`/system/proxy/analytics?since_hours=${sinceHours}`)
    return data.data
  },

  flows: async (opts?: { limit?: number; history?: boolean; host?: string; leaked?: boolean }): Promise<ProxyFlow[]> => {
    const params = new URLSearchParams()
    if (opts?.limit) params.set('limit', String(opts.limit))
    if (opts?.history) params.set('history', 'true')
    if (opts?.host) params.set('host', opts.host)
    if (opts?.leaked) params.set('leaked', 'true')
    const qs = params.toString()
    const { data } = await apiClient.get<ApiResponse<ProxyFlow[]>>(`/system/proxy/flows${qs ? `?${qs}` : ''}`)
    return data.data
  },

  exportCsv: async (opts?: { host?: string; leaked?: boolean }): Promise<Blob> => {
    const params = new URLSearchParams()
    if (opts?.host) params.set('host', opts.host)
    if (opts?.leaked) params.set('leaked', 'true')
    const { data } = await apiClient.get(`/system/proxy/flows/export?${params.toString()}`, { responseType: 'blob' })
    return data as Blob
  },

  flowStats: async (): Promise<ProxyFlowStats> => {
    const { data } = await apiClient.get<ApiResponse<ProxyFlowStats>>('/system/proxy/flows/stats')
    return data.data
  },

  clearFlows: async (): Promise<void> => {
    await apiClient.delete('/system/proxy/flows')
  },
}

File: backend/internal/models/proxy.go
package models

import "time"

// ProxyProfile is one saved egress proxy in the proxy pool. Multiple profiles may
// exist; exactly one is marked IsActive and drives the project-wide egress layer
// (internal/egress). Switching the active profile re-points every outbound client
// at runtime with no restart. URL may embed credentials (socks5://user:pass@host)
// so it is never sent raw to the client — MaskedURL carries the safe form.
type ProxyProfile struct {
	ID   uint   `json:"id" gorm:"primaryKey"`
	Name string `json:"name" gorm:"not null"`
	// URL is the raw proxy URL (http/https/socks5). Never serialized directly.
	URL            string `json:"-"`
	MaskedURL      string `json:"url" gorm:"-"` // scheme://host, credentials stripped — for display
	NoProxy        string `json:"no_proxy"`
	FallbackDirect bool   `json:"fallback_direct" gorm:"default:false"`
	IsActive       bool   `json:"is_active" gorm:"index;default:false"`

	// Lane binds a profile to an egress lane so different traffic classes can use
	// different exits: "default" (project-wide egress), "osint", or "vulnscan".
	// Exactly one profile per lane may be active. Empty is treated as "default".
	Lane string `json:"lane" gorm:"index;default:'default'"`

	// QuotaBytes caps total bytes (in+out) attributed to this profile's label over
	// the retained flow window; 0 = unlimited. Metered residential proxies bill by
	// GB, so this surfaces a soft budget with a UI warning.
	QuotaBytes int64 `json:"quota_bytes" gorm:"default:0"`
	// QuotaHardStop upgrades the quota from advisory to enforced: when usage
	// reaches QuotaBytes the profile is auto-deactivated (its lane drops to direct)
	// so it can't keep spending. Off by default.
	QuotaHardStop bool `json:"quota_hard_stop" gorm:"default:false"`

	// Last health-probe result for this profile (updated by an explicit check or
	// when the profile is active and the background probe runs).
	Healthy   bool       `json:"healthy" gorm:"default:false"`
	LatencyMs int64      `json:"latency_ms" gorm:"default:0"`
	LastError string     `json:"last_error"`
	LastCheck *time.Time `json:"last_check"`

	// Exit identity — the public face traffic presents when routed through this
	// proxy (discovered by an explicit identity probe).
	ExitIP        string     `json:"exit_ip"`
	ExitCountry   string     `json:"exit_country"`
	ExitOrg       string     `json:"exit_org"`
	IsTor         bool       `json:"is_tor" gorm:"default:false"`
	ExitCheckedAt *time.Time `json:"exit_checked_at"`

	// Exit-identity drift: the previously-seen exit IP and a flag raised when a
	// later auto-refresh finds a different IP — an unexpected change can mean the
	// proxy died and traffic fell back, or the exit rotated under you.
	ExitIPPrev    string `json:"exit_ip_prev,omitempty"`
	IdentityDrift bool   `json:"identity_drift" gorm:"default:false"`

	// QuotaUsedBytes / OverQuota are computed (not stored) for the list response.
	QuotaUsedBytes int64 `json:"quota_used_bytes" gorm:"-"`
	OverQuota      bool  `json:"over_quota" gorm:"-"`

	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

// ProxyHealthSample is one timestamped health-probe result for a profile, kept as
// a rolling window so the UI can draw an uptime/latency sparkline per proxy.
type ProxyHealthSample struct {
	ID        uint      `json:"id" gorm:"primaryKey"`
	ProfileID uint      `json:"profile_id" gorm:"index"`
	Healthy   bool      `json:"healthy"`
	LatencyMs int64     `json:"latency_ms"`
	CreatedAt time.Time `json:"created_at" gorm:"index"`
}

// ProxyFlow is one recorded outbound request/response through the egress layer —
// the "flow" the Proxy Manager displays. Persisted (rolling window) so history
// survives a restart; a bounded in-memory ring buffer serves the live view.
type ProxyFlow struct {
	ID        uint      `json:"id" gorm:"primaryKey"`
	CreatedAt time.Time `json:"created_at" gorm:"index"`

	ProxyLabel string `json:"proxy_label"` // active profile name, or "direct"
	ViaProxy   bool   `json:"via_proxy"`
	Leaked     bool   `json:"leaked" gorm:"index"` // went direct while a proxy was active
	Source     string `json:"source"`              // originating module (osint, app, …)

	Method      string `json:"method"`
	Scheme      string `json:"scheme"`
	Host        string `json:"host" gorm:"index"`
	URL         string `json:"url"`
	Status      int    `json:"status"`
	ContentType string `json:"content_type"`
	TLSVersion  string `json:"tls_version"`
	BytesOut    int64  `json:"bytes_out"`
	BytesIn     int64  `json:"bytes_in"`
	DurationMs  int64  `json:"duration_ms"`
	DNSMs       int64  `json:"dns_ms"`
	ConnectMs   int64  `json:"connect_ms"`
	TLSMs       int64  `json:"tls_ms"`
	TTFBMs      int64  `json:"ttfb_ms"`
	Error       string `json:"error"`
}

// ProxyPoolSetting is a singleton row (id = 1) holding pool-automation config:
//   - manual:   the operator switches proxies by hand (default).
//   - failover: if the active proxy's health probe fails, auto-switch to the next
//     healthy profile.
//   - rotate:   round-robin the active proxy across healthy profiles every
//     IntervalSec seconds (rotating exit identity).
type ProxyPoolSetting struct {
	ID          uint   `json:"id" gorm:"primaryKey"`
	Mode        string `json:"mode" gorm:"default:manual"`
	IntervalSec int    `json:"interval_sec" gorm:"default:300"`
	// KillSwitch, when on, HARD-BLOCKS all project-wide egress whenever no healthy
	// proxy is available (none configured, or the active one is confirmed down and
	// fall-back-to-direct is off) — fail-closed so recon can never leak the real
	// IP the moment anonymity drops.
	KillSwitch bool      `json:"kill_switch" gorm:"default:false"`
	UpdatedAt  time.Time `json:"updated_at"`
}
